crates/outbound-networking/src/lib.rs

```diff
@@ -1,122 +1,251 @@
     } else {
         default
     };
 
     if !is_allowed {
-        terminal::warn!("A component tried to make a request to non-allowed address '{address}'.");
-        let (host, port) = (address.host(), address.port());
+        terminal::warn!("A component tried to make a request to non-allowed url '{url}'.");
+        let (scheme, host, port) = (url.scheme, url.host, url.port);
         let msg = if let Some(port) = port {
-            format!("`allowed_outbound_hosts = [\"{host}:{port}\"]`")
+            format!("`allowed_outbound_hosts = [\"{scheme}://{host}:{port}\"]`")
         } else {
-            format!("`allowed_outbound_hosts = [\"{host}:$PORT\"]` (where $PORT is the correct port number)")
+            format!("`allowed_outbound_hosts = [\"{scheme}://{host}:$PORT\"]` (where $PORT is the correct port number)")
         };
         eprintln!("To allow requests, add {msg} to the manifest component section.");
     }
     is_allowed
 }
 
 /// An address is a url-like string that contains a host, a port, and an optional scheme
-#[derive(Debug)]
-struct Address {
-    inner: Url,
+#[derive(Eq, Debug, Clone)]
+pub struct AllowedHostConfig {
     original: String,
-    has_scheme: bool,
+    scheme: SchemeConfig,
+    host: HostConfig,
+    port: PortConfig,
 }
 
-impl Address {
+impl AllowedHostConfig {
     /// Try to parse the address.
     ///
     /// If the parsing fails, the address is prepended with the scheme and parsing
     /// is tried again.
-    pub fn parse(url: &str, scheme: Option<&str>) -> anyhow::Result<Self> {
-        let mut has_scheme = true;
-        let mut parsed = match Url::parse(url) {
-            Ok(url) if url.has_host() => Ok(url),
-            first_try => {
-                // Parsing with 'scheme' resolves the ambiguity between 'spin.fermyon.com:80' and 'unix:80'.
-                // Technically according to the spec a valid url *must* contain a scheme. However,
-                // we allow url-like address strings without schemes, and we interpret the first part as the host.
-                let second_try = format!("{}://{url}", scheme.unwrap_or("scheme"))
-                    .as_str()
-                    .try_into()
-                    .context("could not convert into a url");
-                has_scheme = scheme.is_some();
-                match (second_try, first_try.map_err(|e| e.into())) {
-                    (Ok(u), _) => Ok(u),
-                    // Return an error preferring the error from the first attempt if present
-                    (_, Err(e)) | (Err(e), _) => Err(e),
+    pub fn parse(url: impl Into<String>) -> anyhow::Result<Self> {
+        let original = url.into();
+        let url = original.trim();
+        let scheme_end = url.find("://").with_context(|| {
+            format!("{url:?} does not contain a scheme (e.g., 'http://' or '*://'")
+        })?;
+        let scheme = &url[..scheme_end];
+        let host_and_port = &url[(scheme_end + 3)..];
```

Comment:
Nice convenience method `split_once` for this:
```suggestion
        let (scheme, host_and_port) = url.split_once("://").with_context(|| {
            format!("{url:?} does not contain a scheme (e.g., 'http://' or '*://'")
        })?;
```